core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/validation/ParcelUrlProvider.java

```diff
@@ -0,0 +1,99 @@
+package com.sequenceiq.cloudbreak.service.upgrade.validation;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.common.StackType;
+import com.sequenceiq.cloudbreak.common.exception.BadRequestException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageCatalogException;
+import com.sequenceiq.cloudbreak.core.CloudbreakImageNotFoundException;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.ClusterComponent;
+import com.sequenceiq.cloudbreak.service.image.ImageCatalogService;
+import com.sequenceiq.cloudbreak.service.image.StatedImage;
+import com.sequenceiq.cloudbreak.service.parcel.ParcelService;
+
+@Component
+class ParcelUrlProvider {
+
+    @Inject
+    private ParcelService parcelService;
+
+    @Inject
+    private ImageCatalogService imageCatalogService;
+
+    Set<String> getRequiredParcelsFromImage(String imageCatalogUrl, String imageCatalogName, String imageId, Stack stack) {
+        StatedImage targetImage = getTargetImage(imageCatalogUrl, imageCatalogName, imageId);
+        String cdhRepoUrl = getCdhParcelUrl(targetImage);
+        return StackType.DATALAKE.equals(stack.getType()) ? Collections.singleton(cdhRepoUrl) : getAllParcel(cdhRepoUrl, targetImage, stack);
+    }
+
+    private StatedImage getTargetImage(String imageCatalogUrl, String imageCatalogName, String imageId) {
+        try {
+            return imageCatalogService.getImage(imageCatalogUrl, imageCatalogName, imageId);
+        } catch (CloudbreakImageNotFoundException | CloudbreakImageCatalogException e) {
+            throw new BadRequestException(e.getMessage(), e);
+        }
+    }
+
+    private String getCdhParcelUrl(StatedImage targetImage) {
+        Map<String, String> stack = targetImage.getImage().getStackDetails().getRepo().getStack();
+        return stack.get("redhat7").concat("CDH-").concat(stack.get("repository-version")).concat("-el7.parcel");
```

Comment:
This seems to be fragile. Could you pls check for null of stack.get("redhat7") and fetch stack.repoid in the middle part?